src/inmanta/ast/statements/generator.py

```diff
@@ -268,7 +267,151 @@ def nested_blocks(self) -> Iterator["BasicBlock"]:
         yield self.module
 
 
-class If(ExpressionStatement):
+# TODO: include general flow in docstring (here or requires_emit?): requires_emit -> resume -> execute
+class ListComprehension(RawResumer, ExpressionStatement):
+    """
+    A list comprehension expression, e.g. `["hello {{world}}" for world in worlds]`.
+    """
+
+    __slots__ = ("loop_var", "value_expression", "iterable")
+
+    def __init__(self, value_expression: ExpressionStatement, loop_var: LocatableString, iterable: ExpressionStatement) -> None:
+        super().__init__()
+        self.value_expression: ExpressionStatement = value_expression
+        self.loop_var: LocatableString = loop_var
+        self.iterable: ExpressionStatement = iterable
+        # TODO: extend anchors?
+
+    def normalize(self, *, lhs_attribute: Optional[AttributeAssignmentLHS] = None) -> None:
+        self.value_expression.normalize(lhs_attribute=lhs_attribute)
+        self.iterable.normalize()
+
+    def requires(self) -> list[str]:
+        # exclude loop var, unless it shadows an occurrence in iterable
+        return list(set(self.iterable.requires()) - {self.loop_var} + set(self.value_expression.requires()))
+
+    def requires_emit(
+        self, resolver: Resolver, queue: QueueScheduler, *, lhs: Optional[ResultCollector[object]] = None
+    ) -> dict[object, VariableABC]:
+        """
+        Sets up gradual execution of the list comprehension. Additionally sets up a resumer for when the iterable is complete.
+        Returns as requires the gradual helper's result variable, which will be frozen by the resumer.
+
+        Flow: requires_emit -> gradual execution -> resume -> execute
+        """
+        base_requires: dict[object, VariableABC] = super().requires_emit(resolver, queue)
```

Comment:
this does nothing?

Reply:
It might: the super class emits eager promises. In this case I'm not sure if there are any but I think it's good practice to call it anyway.